go/downloader/verifying-downloader.go

```diff
@@ -0,0 +1,73 @@
+package downloader
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"os"
+
+	"github.com/golang/glog"
+	"github.com/vbauerster/mpb/v5"
+
+	"github.com/jcjones/ct-mapreduce/storage"
+)
+
+type DownloadVerifier interface {
+	IsValid(path string) error
+}
+
+/*
+ * Returns: Boolean of whether the data at finalPath is now valid, and any error. It is possible
+ * that err != nil and yet finalPath is valid, so callers should rely on the boolean and merely
+ * log the error as needed.
+ */
+func DownloadAndVerifyFileSync(ctx context.Context, verifyFunc DownloadVerifier, auditor DownloadAuditor,
+	issuer storage.Issuer, display *mpb.Progress, crlUrl url.URL, finalPath string, maxRetries uint) (bool, error) {
+	dlTracer := NewDownloadTracer()
+	auditCtx := dlTracer.Configure(ctx)
+
+	tmpPath := fmt.Sprintf("%s.tmp", finalPath)
+	defer func() {
+		removeErr := os.Remove(tmpPath)
+		if removeErr != nil && !os.IsNotExist(removeErr) {
+			glog.Warningf("[%s] Failed to remove invalid tmp file %s: %s", issuer.ID(), tmpPath, removeErr)
+		}
+	}()
+
+	dlErr := DownloadFileSync(auditCtx, display, crlUrl, tmpPath, maxRetries)
+	if dlErr != nil {
+		auditor.FailedDownload(issuer, &crlUrl, dlTracer, dlErr)
+		glog.Warningf("[%s] Failed to download from %s to tmp file %s: %s", issuer.ID(), crlUrl.String(), tmpPath, dlErr)
+
+		existingValidErr := verifyFunc.IsValid(finalPath)
+		if existingValidErr == nil {
+			// The existing file at finalPath is OK.
+			return true, dlErr
+		}
```

Comment:
Should we call `auditor.FailedVerifyPath(issuer, finalPath, existingValidErr)` here?

Reply:
Yes, though I'm refactoring to avoid messing that up :) Good catch